### ISSUE TITLE
CultureInfo is not returning proper system language on Mac Big Sur (Version 11.6) M1 (ARM64)

### ISSUE THREAD

**Issue by VadimPeczynski** [dotnet/runtime#73237]
### Description

I want to get operating system language on Mac Big Sur with .NET 6 and .NET 7 preview 6, but it is always returning en. Tested for de, pt and pl

### Reproduction Steps

Can be reproduced with this code snippet:

using System;
using System.Globalization;

Console.WriteLine($"Language UI Culture: {CultureInfo.CurrentUICulture.TwoLetterISOLanguageName}");
Console.WriteLine($"Language Current Culture: {CultureInfo.CurrentCulture.TwoLetterISOLanguageName}");
Console.WriteLine($"Language Installed Culture: {CultureInfo.InstalledUICulture.TwoLetterISOLanguageName}");

### Expected behavior

Language UI Culture: de
Language Current Culture: de
Language Installed Culture: de

### Actual behavior

Language UI Culture: en
Language Current Culture: en
Language Installed Culture: en

### Regression?

It is not working in .NET 5, .NET 6 and .NET 7 preview 6

**Comment by tarekgh** [comment]
@steveisok this is the same issue I believe you have fixed but I am not sure if your fix was in P6? Could you please have a look?

This is not expected to work on .NET 5.0 not .NET 6.0.

**Comment by steveisok** [comment]
> @steveisok this is the same issue I believe you have fixed but I am not sure if your fix was in P6? Could you please have a look? Also, please triage this issue.
> 
> This is not expected to work on .NET 5.0 not .NET 6.0.

For desktop, the change I made was a backstop in the event your `LANG` environment variable was not set.  On OSX, it should always be set and change when you alter your preferred language and region.  

@VadimPeczynski can you confirm your `LANG` environment variable is set to the language / region you want?

**Comment by VadimPeczynski** [comment]
@steveisok @tarekgh thank you for your response. 
Yes, lang variable is set to the proper language e.g. for **pl**:

> LANG="pl_PL.UTF-8"

but for **de** I'm getting:

> LANG=""

**Comment by steveisok** [comment]
Strange, I would have expected osx to set that for you when you change languages.  Does it ask you to reboot?  I'll have to see if I can get a Big Sur machine to test on.

**Comment by VadimPeczynski** [comment]
Yes, I have tried it with automatic restart (just after language change)
and manuał restart (reject restart and did it from main menu). Weirdly I'm
getting only value for polish language (default installed) and not for
languages installed later. I can also check this on Mac Monterey after
system update

śr., 3 sie 2022, 22:41 użytkownik Steve Pfister ***@***.***>
napisał:

**Comment by cmoaciopm** [comment]
@tarekgh I have met the same issue as well. Is there any plan to fix this issue? Thanks.

**Comment by tarekgh** [comment]
@steveisok could you please help with @cmoaciopm question https://github.com/dotnet/runtime/issues/73237#issuecomment-1956258778?

**Comment by jane-mao** [comment]
I have met the same issue as well.
The issue seems to be related to non-standard culture combinations that Mac OS supports. If any of the standard combinations are used (examples: en-US, en-GB, fr-FR, fr-CA, nl-NL) the issue seemingly does not occur.

As soon as I start mixing the language and region, e.g. ja-CN, ko-CN, fr-CN, then the issue occurs, and I got the empty LANG.

**Comment by tarekgh** [comment]
@steveisok please help or redirect as needed. Thanks!

**Comment by steveisok** [comment]
> I have met the same issue as well. The issue seems to be related to non-standard culture combinations that Mac OS supports. If any of the standard combinations are used (examples: en-US, en-GB, fr-FR, fr-CA, nl-NL) the issue seemingly does not occur.
> 
> As soon as I start mixing the language and region, e.g. ja-CN, ko-CN, fr-CN, then the issue occurs, the `Thread.CurrentThread.CurrentCulture` and `Thread.CurrentThread.CurrentUICulture` were `en-CN`, and I got the empty LANG.

The empty `LANG` appears to be the problem as that's something ICU reads directly. We call `uloc_getDefault` first and if it returns us anything other than `en_US_POSIX`, we'll take it. In the case of `en_US_POSIX`, on apple platforms we'll ask the `NSLocale` api's what it thinks the value is. 

@tarekgh one way to fix this would be to always defer to the `NSLocale` api's. That happens on iOS already, but would be a change on desktop. Thoughts?

**Comment by tarekgh** [comment]
> one way to fix this would be to always defer to the NSLocale api's. That happens on iOS already, but would be a change on desktop. Thoughts?

If calling `NSLocale` APIs already I expect you'll get the correct set language on MAC. Why this is not happening? I assume at that time you are not going to the path calling `uloc_getDefault` anyway. right?

**Comment by tarekgh** [comment]
oh, sorry, this is done on iOS. I think it make sense to do it on Mac too. I think this should work. I saw some other users complained about same issue before.
